### PR TITLE
Fix race condition on qrerun and qrun with end job hook

### PR DESCRIPTION
**Pull request by riyazhakki**
#### Describe Bug or Feature

job stays in substate 41 after qrerun in epilogue script and a single rerun with an execjob_end hook leads to run_count 3 and intermittently Exit_Status=-3

Cause:
Race condition occur on force qrerun and qrun of same job, while asynchronous hook (exec_job_end) is running in the background.

#### Describe Your Change

Allowing job to rerun only when it's discard request is acknowledged

#### Attach Test and Valgrind Logs/Output

[TestPbsExecjobEnd.ptl.log](https://github.com/PBSPro/pbspro/files/3983291/TestPbsExecjobEnd.ptl.log)
[TestPbsExecjobEnd.ptl.log](https://github.com/PBSPro/pbspro/files/3974527/TestPbsExecjobEnd.ptl.log)
[valgrind.log](https://github.com/PBSPro/pbspro/files/3974528/valgrind.log)

### CODE REVIEW

**Comment by riyazhakki** [comment]
Hi All,
This PR is up for the review.

File: src/include/job.h [hunk 1]
```diff
@@ -853,14 +855,15 @@ typedef struct	infoent {
 #define IM_RESTART		16
 #define IM_DELETE_JOB		17
 #define IM_REQUEUE		18
-#define	IM_DELETE_JOB_REPLY	19
+/* skip 19 to avoid confilict with IS_DISCARD_JOB */
```
Comment (arungrover) [inline src/include/job.h]: Why is this change needed? One is inter-mom message and the other is inter-server message. How can those two conflict?

File: src/resmom/mom_hook_func.c [hunk 2]
```diff
@@ -3424,7 +3424,32 @@ void reply_hook_bg(job *pjob)
 	struct	batch_request *preq = pjob->ji_preq;
 #endif
 
-	if (pjob->ji_qs.ji_svrflags & JOB_SVFLG_HERE) { /*MS*/
+	if (pjob->ji_hook_running_bg_on == IS_DISCARD_JOB) {
```
Comment (arungrover) [inline src/resmom/mom_hook_func.c]: I do not understand the purpose of ji_hook_running_bg_on. It stores inter-server/inter-mom/batch-request/batch-request+errno in it. These are all independent things, do you know why this one variable is used to store all of these things?

File: src/server/array_func.c [hunk 3]
```diff
@@ -535,6 +535,24 @@ update_subjob_state(job *pjob, int newstate)
...
+ * @param[in]	parent - pointer to the parent job
+ * @param[in]	iindx - first number of range
+ *
+ * @return	status
+ * @retval	-1	-  error
+ */
+int
+get_subjob_discarding(job *parent, int iindx)
+{
+	if (iindx == -1)
```
Comment (arungrover) [inline src/server/array_func.c]: Is there a scenario that iindx can be -1? If not then this function isn't really needed because caller can just access discarding variable of the subjob.

File: src/server/node_manager.c [hunk 4]
```diff
@@ -840,8 +846,14 @@ post_discard_job(job *pjob, mominfo_t *pmom, int newstate)
 		return;
 	}
 
-	if (pjob->ji_qs.ji_substate == JOB_SUBSTATE_RERUN3) {
-		static char ndreque[] = "Job requeued, execution node %s down";
+	if (pjob->ji_qs.ji_substate == JOB_SUBSTATE_RERUN3 || pjob->ji_discarding) {
+		
+		static char *ndreque;
```
Comment (arungrover) [inline src/server/node_manager.c]: why is this static?

File: src/server/req_runjob.c [hunk 5]
```diff
@@ -377,6 +381,9 @@ req_runjob(struct batch_request *preq)
 			/* job already running */
 			req_reject(PBSE_BADSTATE, 0, preq);
 			return;
+		} else if (get_subjob_discarding(parent, i) == 1) {
```
Comment (arungrover) [inline src/server/req_runjob.c]: Why are you passing in 'i' (state of the subjob). Shouldn't you be passing offset to this function? I'd also argue if the function is even needed.

File: src/include/job.h (hunk 1 above)
Comment (riyazhakki) [reply src/include/job.h]: @arungrover Thank you for reviewing and providing the comments.

> @riyazhakki Can you please explain a little more about your change and about what happens in this race condition.

During execution of the epilogue hook if the user hits the 'rerun -W force job_id' then the server sends IS_DISCARD_JOB to mom (mom kills the job and starts the exec_jobend hook in the background) and force queue at #L469, in this case it get scheduled to the same exec_vnode where the previous version was running. While the mom starts running hook in the background at the same time it receives the request to rerun the next version of same job, as mom is not blocked for running exec_jobend (background) hook.
When this job version lands on mom, req_commit() will change its state to 41 and start_exec function will abort it, as it sees its previous version is not yet cleaned up. Later, after execution of execjob_end hook (of old job version) mom will purge the job structure and on the server it remains in 41 state and similarly it happens with qrun also.
 

> Why is this change needed? One is inter-mom message and the other is inter-server message. How can those two conflict?

Sister node can receive both messages IM_DELETE_JOB_REPLY and IS_DISCARD_JOB (in case of node fail requeue), and replying to these request is done in a function reply_hook_bg(), therefore the above change is required.

File: src/resmom/mom_hook_func.c (hunk 2 above)
Comment (riyazhakki) [reply src/resmom/mom_hook_func.c]: ji_hook_running_bg_on is used indicate the hook is running in the background and it stores inter-server/inter-mom/batch-request/batch-request+errno in it to send appropriate reply from the deferred function (reply_bg_hook).

File: src/server/array_func.c (hunk 3 above)
Comment (riyazhakki) [reply src/server/array_func.c]: Yes, function subjob_index_to_offset can make it -1.

File: src/server/node_manager.c (hunk 4 above)
Comment (riyazhakki) [reply src/server/node_manager.c]: I have retained it as before.

File: src/server/req_runjob.c (hunk 5 above)
Comment (riyazhakki) [reply src/server/req_runjob.c]: Yes, this should be offset, will change it.

File: src/resmom/mom_hook_func.c (hunk 2 above)
Comment (arungrover) [reply src/resmom/mom_hook_func.c]: This seems incorrect. A variable should not store 2 or more distinct things in it. Like in the case statement you check for "PBS_BATCH_DeleteJob + PBSE_SISCOMM" this will be same as "PBS_BATCH_Commit + PBSE_SVRDOWN". While the example above may not be practically possible but it is a possibility that at some point someone might hit this issue and it then it will be very difficult to diagnose.

Comment (riyazhakki) [reply src/resmom/mom_hook_func.c]: I see even it is two distinct values, will be unique for the event so there is little chance of getting into the issue.
The reason for holding these values is, when MS receives PBS_BATCH_DeleteJob request, it starts execution of the hook in the background and at the same time it sends the IM_DELETE_JOB_REPLY to all its sisters, if it is unable to send this request to one or more sisters, we need to reply the batch request with PBSE_SISCOMM error and this reply is done in the deferred function of hook running in the background, to carry this error value from function req_deletejob() to reply_hook_bg() we are using the ji_hook_running_bg_on variable.

**Comment by riyazhakki** [comment]
[TestPbsExecjobEnd.ptl.log](https://github.com/PBSPro/pbspro/files/4030927/TestPbsExecjobEnd.ptl.log)

File: src/resmom/mom_hook_func.c (hunk 2 above)
Comment (arungrover) [reply src/resmom/mom_hook_func.c]: I hear what you are saying, but I don't like the fact that there is already a clash of values stored in this variable which is causing you to change the inter-mom request command IDs. This means that older moms and newer moms will not understand the messaged being passed between them (although, this may not be supported but it is possible, especially during upgrades)
I'd suggest that instead of storing existing batch request IDs, IM/IS command IDs, error number (or combinations of all three) you create a separate enum and add all possible values this variable is currently used for in that enum and then you don't have to change the Inter-mom command.

**Comment by riyazhakki** [comment]
[TestPbsExecjobEnd.ptl.log](https://github.com/PBSPro/pbspro/files/4045732/TestPbsExecjobEnd.ptl.log)
